banzai_nres/traces.py

```diff
@@ -23,48 +23,49 @@
 MIN_TRACE_HALF_WIDTH = 50
 
 # The final trace will be +- this from the center in the y-direction
-TRACE_HALF_HEIGHT = 6
+TRACE_HALF_HEIGHT = 7
 
 
 def find_y_center(y, indices, weights=None):
     if weights is None:
         weights = np.ones(y.shape, dtype=y.dtype)
     centers = (y * indices * weights).sum(axis=0) / (indices * weights).sum(axis=0)
     # Errors are sqrt sum of the weights
-    errors = np.sqrt((indices * weights).sum(axis=0))
+    errors = 1.0 / np.sqrt((indices * weights).sum(axis=0))
     return centers, errors
 
 
-def refine_traces(traces, image, weights=None):
-    x2d, y2d = np.meshgrid(np.arange(traces.shape[1]), np.arange(traces.shape[0]))
+def refine_traces(image, weights=None):
+    x2d, y2d = np.meshgrid(np.arange(image.traces.shape[1]), np.arange(image.traces.shape[0]))
     # For each label
-    for i in range(1, np.max(traces) + 1):
+    for i in range(1, np.max(image.traces) + 1):
         # TODO: Debug divide by zero in weights
-        y_center, y_center_errors = find_y_center(y2d, traces == i, weights=weights)
+        y_center, y_center_errors = find_y_center(y2d, image.traces == i, weights=weights)
         # Refit the centroids to reject cosmic rays etc, but only evaluate where the S/N is good
-        x_center = np.arange(min(x2d[traces == i]), max(x2d[traces == i]) + 1, dtype=np.float)
+        x_center = np.arange(min(x2d[image.traces == i]), max(x2d[image.traces == i]) + 1, dtype=np.float)
         logger.info(f'Fitting a polynomial to order {i}', image=image)
         # we chose order 5 based on visually inspecting the residuals between the trace centers and the model fit centers
         # TODO we need to verify that an order 5 polynomial fit is the best thing to do.
         best_fit = fit_polynomial(y_center[x_center.astype(int)], y_center_errors[x_center.astype(int)], x=x_center, order=5)
         # TODO: extrapolate fit and go out to pixels that have a cumulative S/N = 10
         y_center = best_fit(x_center)
 
+        y_center = np.round(y_center).astype(int)
         # Pad y_center with zeros so that it has the same dimension as y2d
         padded_y_center = np.zeros(y2d.shape[1])
         padded_y_center[int(min(x_center)):int(max(x_center) + 1)] = y_center[:]
         pixels_to_label = np.logical_and(np.logical_and(x2d >= min(x_center), x2d <= max(x_center)),
                                          np.abs(y2d - padded_y_center) <= TRACE_HALF_HEIGHT)
         # Reset the previously marked traces to 0. Then mark the newly measured traces.
```

Comment:
I would rather have TRACE_HALF_HEIGHT be a property of the class, so that it can be changed in the unit test that is called (and any other tests that we might end up writing).